src/System.Net.HttpListener/src/System/Net/Windows/AsyncRequestContext.cs

```diff
@@ -23,19 +39,28 @@ internal AsyncRequestContext(ThreadPoolBoundHandle boundHandle, ListenerAsyncRes
         private Interop.HttpApi.HTTP_REQUEST* Allocate(ThreadPoolBoundHandle boundHandle, uint size)
         {
             uint newSize = size != 0 ? size : RequestBuffer == null ? 4096 : Size;
-            if (_nativeOverlapped != null && newSize != RequestBuffer.Length)
+            if (_nativeOverlapped != null)
             {
+#if DEBUG
+                DebugRefCountReleaseNativeOverlapped();
+#endif
+
                 NativeOverlapped* nativeOverlapped = _nativeOverlapped;
                 _nativeOverlapped = null;
                 _boundHandle.FreeNativeOverlapped(nativeOverlapped);
             }
+
             if (_nativeOverlapped == null)
             {
+#if DEBUG
+                DebugRefCountAllocNativeOverlapped();
+#endif
                 SetBuffer(checked((int)newSize));
                 _boundHandle = boundHandle;
                 _nativeOverlapped = boundHandle.AllocateNativeOverlapped(ListenerAsyncResult.IOCallback, state: _result, pinData: RequestBuffer);
                 return (Interop.HttpApi.HTTP_REQUEST*)Marshal.UnsafeAddrOfPinnedArrayElement(RequestBuffer, 0);
             }
+
             return RequestBlob;
```

Comment:
Good point: we need to remove this line entirely and the last if statement.

The 2 callers of `AsyncRequestContext` set `RequestBlob` to be equal to what `AsyncRequestContext` returns. The previous optimization assumed that the struct can be reused and so no blob initialization was necessary (it was being initialized to itself).